### PR TITLE
added filters and sorting for products

### PR DESCRIPTION
**Pull request by JantuDeb**
### Issues #8 #6 
# Filters
- Filter by price
- Filter by rating
- Filter by categories

# Sorting
- Sort by new products
- Sort by price (low to high)
- Sort by price (high to low)
image

### CODE REVIEW

File: src/utils/filters.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+export const composeFilterFunc =
+  (state, ...functions) =>
+  (products) =>
+    functions.reduce((acc, curr) => {
+      return curr(state, acc);
+    }, products);
+
+export const sortProduct = (state, products) => {
+  console.log(state.sortBy.name);
```
Comment (rkapoor10) [inline src/utils/filters.js:11]: remove console.log

File: src/components/products/Product.jsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+          <div className="flex center">
+            <div className="bg-green px-2 text-white flex center radius-sm">
+              {product.rating} <AiFillStar />
+            </div>
+            <p className="text-center text-red m-0 px-2">
+              ₹{getDiscountedPrice(product.price, product.discount)}
+              <s className="text-gray">₹{product.price}</s>
+              {product.discount * 100}% off
+            </p>
+          </div>
```
Comment (mariyasada) [inline src/components/products/Product.jsx:37]: liked that you created separate functions

File: src/utils/utils.js [hunk 3]
```diff
@@ -0,0 +1,2 @@
+export const getDiscountedPrice = (price, discount) =>
```
Comment (rkapoor10) [inline src/utils/utils.js:1]: liked that you extracted this in a function

File: src/context/products/ProductContext.jsx [hunk 4]
```diff
@@ -1,24 +1,51 @@
...
-  console.log(products);
+  const filterProducts = composeFilterFunc(
+    productState,
+    sortProduct,
+    filterCategory,
+    filterRating,
+    filterPrice
+  )(products);
+
+  console.log(productState);
```
Comment (rkapoor10) [inline src/context/products/ProductContext.jsx:44]: remove console.log

**Review by rkapoor10 — APPROVED** [review]
The code is well structured and the filters functions are well separated. No refetching of data. Looks good to merge.

File: src/context/products/ProductContext.jsx (hunk 4 above)
Comment (mariyasada) [inline src/context/products/ProductContext.jsx:44]: remove console.log

**Review by mariyasada — APPROVED** [review]
I like that you created array of categories and ratings and mapping on it and its good that you created small functions like getDiscountedPrice

**Review by Bhakti199 — APPROVED** [review]
This code looks fine to me.
Add range in between , because at 0 it shows all products.
Overall the code is written well and structured properly.
This PR is good to merge.

**Pull request merged**